compiler/src/dmd/iasmgcc.d

```diff
@@ -175,12 +171,43 @@ bool requireToken(Parser)(Parser p, TOK value)
     return false;
 }
 
+/***********************************
+ * Run semantic analysis on `exp`, resolving it as a compile-time string.
+ * Params:
+ *      sc = scope
+ *      exp = Expression which expected as a string
+ *      s = What the string is expected for, used in error diagnostic
+ * Returns:
+ *      StringExp or ErrorExp
+ */
+Expression semanticAsmString(Scope* sc, Expression exp, const char *s)
+{
+    import dmd.dcast : implicitCastTo;
+    import dmd.dsymbolsem : resolveAliasThis;
+    import dmd.mtype : isAggregate, Type;
+
+    exp = expressionSemantic(exp, sc);
+
+    // Resolve `alias this` if we were given a struct literal.
+    if (auto ad = isAggregate(exp.type))
+    {
+        if (ad.aliasthis && ad.type && !ad.type.isTypeError())
+            exp = resolveAliasThis(sc, exp);
+    }
+
+    // Evaluate the expression as a string now or error trying.
+    if (auto se = semanticString(sc, exp, s))
+        exp = implicitCastTo(exp, sc, Type.tstring);
+
+    return exp;
+}
+
 /***********************************
  * Parse an expression that evaluates to a string.
  * Grammar:
  *      | AsmStringExpr:
  *      |     StringLiteral
- *      |     ( AssignExpression )
+ *      |     ( ConditionalExpression )
```

Comment:
It makes no sense allowing AssignExpression for a constant expression.  Tighten it to the next expression in the grammar.